{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "b744c1ef-1280-4d26-a01a-8bc44b5b20a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from my_module import mstools\n",
    "from my_module import forward_trajectory as fwd\n",
    "from my_module import mbslib\n",
    "import csv\n",
    "import pprint"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "0b720264-355a-4a63-8d4d-17d0e5b23960",
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_mbs(nsam, per_site_theta, per_site_rho,\n",
    "            lsites, selpos,\n",
    "            n_trajectory, nrep_per_traj,\n",
    "            path_to_mbs_output, path_to_traj):\n",
    "    '''\n",
    "    Args:\n",
    "        nsam,\n",
    "        per_site_theta,\n",
    "        per_site_rho,\n",
    "        lsites,\n",
    "        selpos,\n",
    "        n_trajectory (int): \n",
    "        nrep_per_traj (int): \n",
    "        path_to_mbs_output (str) : path to mbs output files (w/o extentions)\n",
    "        path_to_traj (str) : path to trajectory files (w/o extentions)\n",
    "    '''\n",
    "\n",
    "    cmd = f'mbs {nsam} -t {per_site_theta} -r {per_site_rho} '\n",
    "    cmd += f'-s {lsites} {selpos} '\n",
    "    cmd += f'-f {n_trajectory} {nrep_per_traj} {path_to_traj} '\n",
    "    cmd += f'> {path_to_mbs_output}'\n",
    "\n",
    "    mbslib.run_command(cmd)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "8e57b9c4-beac-4aab-9527-a3048ebfb6c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_trajectoryfiles_forward(N0, generation, demography_in_year, t_mutation_in_year, s, h, resolution,\n",
    "                         n_trajectory, path_to_traj):\n",
    "    '''\n",
    "    generate trajectory files\n",
    "\n",
    "    Args:\n",
    "        N0 (int):\n",
    "        generation (int): generation time, years/generation\n",
    "        demography_in_year (list): demographic history/\n",
    "        t_mutation_in_year (float): time when mutation arises, in year\n",
    "        s,\n",
    "        h,\n",
    "        resolution,\n",
    "        n_trajectory (int): number of trajectories\n",
    "        path_to_traj (str) : path to trajectory files (w/o extentions)\n",
    "\n",
    "    '''\n",
    "    for i in range(n_trajectory):\n",
    "\n",
    "        filename = f'{path_to_traj}_{i}.dat'\n",
    "\n",
    "        # generate trajectory\n",
    "        trajectory = fwd.mbs_input(t_mutation_in_year,\n",
    "                                   demography_in_year,\n",
    "                                   s, h,\n",
    "                                   generation, N0, resolution,\n",
    "                                   'NOTLOST')\n",
    "\n",
    "        # save\n",
    "        with open(filename, 'w') as f:\n",
    "            writer = csv.writer(f, delimiter='\\t')\n",
    "            for freq in trajectory:\n",
    "                writer.writerow(freq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "604123ea-0d98-4c61-8c84-fe170cf5df59",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parameter_sets_forward(t_mutation_in_year, sel_advantages, nrep):\n",
    "    params = dict()\n",
    "\n",
    "    params['N0'] = 5000\n",
    "    params['generation'] = 20\n",
    "    params['demography_in_year'] = [[0, 100 * params['N0'] * params['generation'], params['N0']]]\n",
    "\n",
    "    # selection coefficients\n",
    "    params['s'] = 0\n",
    "    params['h'] = 0.5  # <--- co-dominance\n",
    "    params['resolution'] = 100\n",
    "\n",
    "    # number of trajectory\n",
    "    params['n_trajectory'] = nrep\n",
    "    # coalescent simulation per trajectory\n",
    "    params['nrep_per_traj'] = 1\n",
    "\n",
    "    # number of chromosome\n",
    "    params['nsam'] = 120\n",
    "    # length of sequence\n",
    "    params['lsites'] = 10000\n",
    "    # position of target site\n",
    "    params['selpos'] = params['lsites']/2\n",
    "\n",
    "    # mutation rate per site per generation\n",
    "    params['per_site_theta'] = 1.0 * 10 ** (-8) * 4 * params['N0']\n",
    "    # recombination rate per site per generation\n",
    "    params['per_site_rho'] = 1.0 * 10 ** (-8) * 4 * params['N0']\n",
    "\n",
    "\n",
    "    params_list = list()\n",
    "    for s in sel_advantages:\n",
    "        params['s'] = s\n",
    "        for t in t_mutation_in_year:\n",
    "            params['t_mutation_in_year'] = t\n",
    "            params_list.append(params.copy())\n",
    "\n",
    "    return params_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "843a775c-18a6-4a83-97c7-e92fb053b9ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_thetapi_S_thetaw_thetal(ms_seq_data, ms_pos_data):\n",
    "    \"\"\" Calculate theta\n",
    "\n",
    "     Args:\n",
    "         ms_seq_data(list(str): ms 0-1 sequence data of one replication\n",
    "         ms_pos_data(list(float)): SNP position data of one replication\n",
    "\n",
    "     Return:\n",
    "         thetapi:\n",
    "         S:\n",
    "         thetaw:\n",
    "         thetal:\n",
    "         thetah:\n",
    "\n",
    "     \"\"\"\n",
    "    #number of sample\n",
    "    nsam = len(ms_seq_data)\n",
    "\n",
    "    #get sequence data in int\n",
    "    int_site_list = [[int(i) for i in list(j)] for j in ms_seq_data]\n",
    "    \n",
    "    #calc theta_l\n",
    "    #calc number of segregating sites\n",
    "    l = sum([sum(i) for i in int_site_list])\n",
    "    #calc average number of segregating sites\n",
    "    thetal = l/(nsam - 1)\n",
    "\n",
    "    #calc theta_pi, theta_h\n",
    "    #calc sum of pairwise differences\n",
    "    k = 0\n",
    "    h = 0\n",
    "    for i in zip(*int_site_list):\n",
    "        der = sum(i)\n",
    "        k += der*(nsam-der)\n",
    "        h += der**2\n",
    "    #clac_thetapi/h\n",
    "    thetapi = k*2/(nsam*(nsam-1))\n",
    "    thetah = h*2/(nsam*(nsam-1))\n",
    "\n",
    "    #calc theta_w\n",
    "    #calc number of segregating sites\n",
    "    S = len(ms_pos_data)\n",
    "    #calc theta_w\n",
    "    a = 0\n",
    "    for j in range(1,nsam):\n",
    "        a += 1/j\n",
    "    thetaw = S/a\n",
    "    \n",
    "    return thetapi, S, thetaw, thetal, thetah"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "03f482c0-8598-4964-a964-1e3de84dcb2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_H(thetapi, S, thetaw, thetal, thetah, n):\n",
    "    \"\"\" Calculate normalized version of Fay and Wu's H\n",
    "\n",
    "    Args:\n",
    "        thetapi:\n",
    "        S:\n",
    "        thetaw:\n",
    "        thetal:\n",
    "        thetah:\n",
    "        n: sample size\n",
    "\n",
    "    Return:\n",
    "        H: Fay and Wu H\n",
    "    \"\"\"\n",
    "    #calc variation of H\n",
    "    a = sum([1/i for i in range(1, n)])\n",
    "    b = sum([1/(i**2) for i in range(1, n)])\n",
    "    v = (n-2)*thetaw/(6*(n-1)) + (18*n**2*(3*n+2)*(b+1/n**2)-(88*n**3+9*n**2-13*n+6))*(S*(S-1)/(a**2+b))/(9*n*(n-1)**2)\n",
    "    #return nan if number of segregating sites is too small to calculate H\n",
    "    if v == 0:\n",
    "        return np.nan\n",
    "    else:\n",
    "        H = (thetapi-thetal)/v**(1/2)\n",
    "        return H"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "d54eb463-2010-4b2c-9d69-79df5390c895",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_D(thetapi, S, thetaw, thetal, thetah, n):\n",
    "    \"\"\" Calculate Tajima's D\n",
    "\n",
    "    Args:\n",
    "        thetapi:\n",
    "        S: number of segregating sites\n",
    "        thetaw:\n",
    "        thetal:\n",
    "        thetah:\n",
    "        n: sample size\n",
    "\n",
    "    Return:\n",
    "         D: Tajima's D\n",
    "    \"\"\"\n",
    "    #calc variation of D\n",
    "    a1 = 0\n",
    "    for i in range(1,n):\n",
    "        a1 += 1/i\n",
    "    a2 = 0\n",
    "    for i in range(1, n):\n",
    "        a2 += 1/i**2\n",
    "    b1 = (n + 1)/(3*(n-1))\n",
    "    b2 = 2*(n**2 + n + 3)/(9*n*(n-1))\n",
    "    c1 = b1 - 1/a1\n",
    "    c2 = b2 - (n+2)/(a1*n) + a2/a1**2\n",
    "    e1 = c1/a1\n",
    "    e2 = c2/(a1**2 + a2)\n",
    "    C = (e1*S + e2*S*(S-1))**0.5\n",
    "    if C == 0:\n",
    "        return np.nan\n",
    "    else:\n",
    "        D = (thetapi - thetaw)/C\n",
    "        return D"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "2920cc6f-32a7-4308-bfd1-26fa41327026",
   "metadata": {},
   "outputs": [],
   "source": [
    "def ms(nsam, nreps, theta, r_rate, nsite, filename):\n",
    "    \"\"\" Run ms\n",
    "\n",
    "    Args:\n",
    "        nsam: number of chromosomes\n",
    "        nreps: number of replication\n",
    "        theta: population mutation parameter per region\n",
    "        r_rate: population recombination parameter per region\n",
    "        nsite: length of simulated region\n",
    "    \"\"\"\n",
    "    cmd = \"ms {} {} -t {} -r {} {} > {}\".format(nsam, nreps, theta, r_rate, nsite, filename)\n",
    "    mstools.run_command(cmd)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "307847e3-9494-4210-be24-9a314c7ba067",
   "metadata": {},
   "outputs": [],
   "source": [
    "def base_to_seq(hap, base_pos, base_len):\n",
    "    \"\"\" Produces a string of length 'base_len' with A indicating a fixed base\n",
    "        C indicating an ancestral base at a polymorphic position and G \n",
    "        indicating a derived base. \"\"\"\n",
    "    cur_site = 0\n",
    "    cur_base = 0\n",
    "    seq_parts = list()\n",
    "    while cur_site < len(base_pos):\n",
    "        # fill in the empty space\n",
    "        seq_parts.append('A' * (base_pos[cur_site] - cur_base))\n",
    "        if (hap[cur_site] == '0'):\n",
    "            seq_parts.append('C')\n",
    "        else:\n",
    "            seq_parts.append('G')\n",
    "        cur_base = base_pos[cur_site] + 1\n",
    "        cur_site += 1\n",
    "    if cur_base < base_len:\n",
    "        seq_parts.append('A' * (base_len - cur_base))\n",
    "    return ''.join(seq_parts)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "4f93bd18-1e62-41c7-a966-63fc1be96e45",
   "metadata": {},
   "outputs": [],
   "source": [
    "def ms2fasta(fasta_output, seq):\n",
    "    # \n",
    "    base_pos = [0 + i for i in range(len(seq[0]))]\n",
    "    base_len = len(seq[0])\n",
    "    outgroup_seq = 'C'*base_len\n",
    "    # \n",
    "    with open(fasta_output, \"w\") as f:\n",
    "        # write outgroup seq\n",
    "        f.write('>seq{}'.format('0'))\n",
    "        f.write(\"\\n\")\n",
    "        f.write(outgroup_seq)\n",
    "        f.write(\"\\n\\n\")\n",
    "        # write sample seq\n",
    "        for m, n in enumerate(seq):\n",
    "            f.write('>seq{}'.format(m+1))\n",
    "            f.write(\"\\n\")\n",
    "            line = base_to_seq(n, base_pos, base_len)\n",
    "            f.write(line)\n",
    "            f.write(\"\\n\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "1cbdaae8-b58b-470c-be76-63a223248ffa",
   "metadata": {},
   "outputs": [],
   "source": [
    "def mbs2ms(mbs_input_file, ms_output_file, nsam, n_traj,\n",
    "                 per_site_theta, per_site_rho, selpos, lsites):\n",
    "    # generate file\n",
    "    with open(ms_output_file, 'w') as f:\n",
    "        # convert mbs format to ms format\n",
    "        f.write(\"ms {} {} -t {} -r {} {}\\n\\n\".format(nsam, n_traj,\n",
    "                                                     per_site_theta * lsites,\n",
    "                                                     per_site_rho * lsites, lsites))\n",
    "\n",
    "        # convert into ms format for each line\n",
    "        for m in mbslib.parse_mbs_data(mbs_input_file):\n",
    "            f.write(\"//\\n\")\n",
    "            # write segregating sites\n",
    "            f.write(\"segsites: {}\\n\".format(len(m['pos'])))\n",
    "\n",
    "            # write position\n",
    "            f.write(\"positions: \")\n",
    "            # convert int to str\n",
    "            pos_list = [str(i/lsites) for i in m['pos']]\n",
    "            f.write(\" \".join(pos_list))\n",
    "            f.write(\"\\n\")\n",
    "\n",
    "            # write seq\n",
    "            f.write(\"\\n\".join(m[\"seq\"]))\n",
    "            f.write(\"\\n\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73abaffa-a944-4702-9bdc-c960133eef11",
   "metadata": {},
   "source": [
    "### simulate neutral case "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "eb5e21ce-a3cd-4e9b-b622-8b1ce658040c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# run ms to generate sample data\n",
    "# parameter sets\n",
    "# initial vales\n",
    "nsam = 120\n",
    "nreps = 1\n",
    "theta = 2\n",
    "r = 2\n",
    "lsites = 10000\n",
    "filename = \"results/ms_data.txt\"\n",
    "# run ms \n",
    "ms(nsam, nreps, theta, r, lsites, filename)\n",
    "\n",
    "# convert to fasta format\n",
    "fasta_output = 'sample_data.fas'\n",
    "for m in mstools.parse_ms_data(filename):\n",
    "    seq = m['seq']\n",
    "    ms2fasta(fasta_output, seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "c9825ad5-3035-4cfd-9da7-37b819e260e3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TajiD in-house neutral case: -0.6327086308397178\n",
      "FWH in-house neutral case: 0.18630539898438378\n"
     ]
    }
   ],
   "source": [
    "### calculate H and D\n",
    "for m in mstools.parse_ms_data(filename):\n",
    "    theta = calc_thetapi_S_thetaw_thetal(m['seq'], m['pos'])\n",
    "    TajiD_inhouse = calc_D(*theta, nsam)\n",
    "    FWH_inhouse = calc_H(*theta, nsam)\n",
    "print('TajiD in-house neutral case:', TajiD_inhouse)\n",
    "print('FWH in-house neutral case:', FWH_inhouse)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e85b27b-314d-422d-982f-42a5497c0a39",
   "metadata": {},
   "source": [
    "### simulate selective case"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "6e191a39-a30e-4a97-94e5-e9138cc7a31f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# initial values\n",
    "# in units of year\n",
    "mutation_ages = [10000]\n",
    "# selection\n",
    "sel_advantages = [0.01]\n",
    "# number of trajectory\n",
    "number_of_trajectory = 1\n",
    "# parameters sets\n",
    "params = parameter_sets_forward(mutation_ages, sel_advantages, number_of_trajectory)[0]\n",
    "\n",
    "# run mbs to genrate sample sequences\n",
    "# path to trajectory\n",
    "#path_to_traj = \"results/traj_t{}_s{}\".format(params['t_mutation_in_year'], params['s'])\n",
    "\n",
    "# trajectoryの生成\n",
    "\n",
    "make_trajectoryfiles_forward(params['N0'], params['generation'],\n",
    "                              params['demography_in_year'], params['t_mutation_in_year'],\n",
    "                              params['s'], params['h'], params['resolution'],\n",
    "                              params['n_trajectory'], path_to_traj)\n",
    "                              \n",
    "\n",
    "# mbs output filename\n",
    "path_to_mbs_output = \"sample_data_selective/mbs_nsam{}_t{}_s{}.dat\".format(params['nsam'], params['t_mutation_in_year'], params['s'])\n",
    "\n",
    "# run mbs\n",
    "run_mbs(params['nsam'], params['per_site_theta'], params['per_site_rho'],\n",
    "        params['lsites'], params['selpos'],\n",
    "        params['n_trajectory'], params['nrep_per_traj'],\n",
    "        path_to_mbs_output, path_to_traj)\n",
    "\n",
    "# convert mbs output to ms output\n",
    "ms_output = \"ms_nsam{}_t{}_s{}.txt\".format(params['nsam'], params['t_mutation_in_year'], params['s'])\n",
    "mbs2ms(path_to_mbs_output, ms_output, params['nsam'], params['n_trajectory'], \n",
    "            params['per_site_theta'], params['per_site_rho'], params['selpos'], params['lsites'])\n",
    "\n",
    "# convert ms output to fasta format\n",
    "# convert to fasta format\n",
    "fasta_output = \"fasta_nsam{}_t{}_s{}.fas\".format(params['nsam'], params['t_mutation_in_year'], params['s'])\n",
    "for m in mstools.parse_ms_data(ms_output):\n",
    "    seq = m['seq']\n",
    "    ms2fasta(fasta_output, seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "dce95b8a-37e5-440e-aa43-b8dfb116039c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "000100000001\n",
      "TajiD in-house selective case: 1.0312906226156557\n",
      "FWH in-house selective case: 1.0641892876470052\n"
     ]
    }
   ],
   "source": [
    "# mbs output filename\n",
    "# path_to_mbs_output = \"mbs_nsam{}_t{}_s{}.dat\".format(params['nsam'], params['t_mutation_in_year'], params['s'])\n",
    "path_to_mbs_output = 'sample_data_selective/mbs_nsam120_t10000_s0.01.dat'\n",
    "# in units of year\n",
    "mutation_ages = [10000]\n",
    "# selection\n",
    "sel_advantages = [0.01]\n",
    "# number of trajectory\n",
    "number_of_trajectory = 1\n",
    "# parameters sets\n",
    "params = parameter_sets_forward(mutation_ages, sel_advantages, number_of_trajectory)[0]\n",
    "# convert mbs output to ms output\n",
    "ms_output = \"ms_nsam{}_t{}_s{}.txt\".format(params['nsam'], params['t_mutation_in_year'], params['s'])\n",
    "# calculate stats\n",
    "for m in mbslib.parse_mbs_data(path_to_mbs_output):\n",
    "    print(m['seq'][0])\n",
    "    theta = calc_thetapi_S_thetaw_thetal(m['seq'], m['pos'])\n",
    "    TajiD_inhouse = calc_D(*theta, params['nsam'])\n",
    "    FWH_inhouse = calc_H(*theta, params['nsam'])\n",
    "\n",
    "print('TajiD in-house selective case:', TajiD_inhouse)\n",
    "print('FWH in-house selective case:', FWH_inhouse)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa1187a2-4953-44e8-8e6b-0fc3398cddea",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
